chia/util/config.py

```diff
@@ -41,35 +47,72 @@ def config_path_for_filename(root_path: Path, filename: Union[str, Path]) -> Pat
     return root_path / "config" / filename
 
 
+def get_config_lock(root_path: Path, filename: Union[str, Path]) -> BaseFileLock:
+    lock_path: Path = config_path_for_filename(root_path, filename).with_suffix(".lock")
+    if not lock_path.exists():
+        lock_path.touch(exist_ok=True)
+    return FileLock(lock_path)
+
+
 def save_config(root_path: Path, filename: Union[str, Path], config_data: Any):
+    # This must be called under an acquired config lock
     path: Path = config_path_for_filename(root_path, filename)
-    tmp_path: Path = path.with_suffix("." + str(os.getpid()))
-    with open(tmp_path, "w") as f:
-        yaml.safe_dump(config_data, f)
-    try:
-        os.replace(str(tmp_path), path)
-    except PermissionError:
-        shutil.move(str(tmp_path), str(path))
+    with tempfile.TemporaryDirectory(dir=path.parent) as tmp_dir:
+        tmp_path: Path = Path(tmp_dir) / Path(filename)
+        with open(tmp_path, "w") as f:
+            yaml.safe_dump(config_data, f)
+        try:
+            os.replace(str(tmp_path), path)
+        except PermissionError:
+            shutil.move(str(tmp_path), str(path))
 
 
 def load_config(
     root_path: Path,
     filename: Union[str, Path],
     sub_config: Optional[str] = None,
-    exit_on_error=True,
+    exit_on_error: bool = True,
+    acquire_lock: bool = True,
 ) -> Dict:
+    # This must be called under an acquired config lock, or acquire_lock should be True
+
     path = config_path_for_filename(root_path, filename)
-    if not path.is_file():
-        if not exit_on_error:
-            raise ValueError("Config not found")
-        print(f"can't find {path}")
-        print("** please run `chia init` to migrate or create new config files **")
-        # TODO: fix this hack
-        sys.exit(-1)
-    r = yaml.safe_load(open(path, "r"))
-    if sub_config is not None:
-        r = r.get(sub_config)
-    return r
+
+    config_lock: Optional[BaseFileLock] = None
+    if acquire_lock:
+        config_lock = get_config_lock(root_path, filename)
+        acquired = config_lock.acquire()
+        if not acquired:
+            raise RuntimeError("Not able to acquire lock in load_config")
```

Comment:
I think this is a good case for an exit stack context manager, if the conditional acquisition is needed.  But, filelock specifically says it allows nesting which both means that this is unneeded and that the lock can't be used with async in the body.

Reply:
Silly mistake.  A _single_ lock object can be nested, separate lock objects on the same file cannot.